### PR TITLE
Rename TensorVector to TensorList

### PR DESCRIPTION
**Pull request by klecki**
## Description:
TensorList was replaced by a refactored TensorVector that supports both
the TensorList APIs as well as non-contiguous mode in which in can be
constructed from individual samples. The TensorList was kept as an
alias to TensorVector to limit the code changes.

As the TensorList name is more prelevant across DALI codebase, TensorVector
is renamed into TensorList and the alias is removed, leaving us with just
one batch type.

The tests are merged into one file, taking the ones from original
TensorVector and the ones that were ported to it from TensorList tests.
Old TensorList tests are removed as they were less generic.

Changes are fully automated, both the search&replace as well as code
formatting.

**GitHub doesn't recognize the moves, hence the big number of lines:**
```
tensor_vector.h -> tensor_list.h
tensor_vector.cc -> tensor_list.cc 

tensor_list_test.cc - original contents removed
{tensor_vector_test.cc, tensor_list_as_tensor_vector_test.cc} moved into: tensor_list_test.cc
```

Note: I am sticking to just renaming TensorVector to not obscure it with other changes.
In a followup I will adjust the UnsafeSet/Copy sample API and some docs for them,
as well as clear External Source of some (now) excess members.

### Affected modules and functionalities:
TensorVector, TensorList, tests for those

### Key points relevant for the review:
This is mostly rename.

### Tests:

- [x] Existing tests apply
Old TensorList tests were removed (we had them ported to TensorVector and generalized over contiguous/noncontiguous cases).
The ported tests and TensorVector tests were renamed and moved into tensor_list_test.cc.
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2328

### CODE REVIEW

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
I must sprinkle the tensor_list.h with DLL_PUBLIC (I think). What is curious, the linking problem doesn't happen in my local build.

**Comment by klecki** [comment]
!build

File: dali/pipeline/data/tensor_list.cc [hunk 1]
```diff
@@ -241,7 +241,7 @@ TensorVector<Backend> &TensorVector<Backend>::operator=(TensorVector<Backend> &&
 
 
 template <typename Backend>
-void TensorVector<Backend>::VerifySampleShareCompatibility(DALIDataType type, int sample_dim,
+void TensorList<Backend>::VerifySampleShareCompatibility(DALIDataType type, int sample_dim,
                                                            TensorLayout layout, bool pinned,
                                                            AccessOrder order, int device_id,
                                                            const std::string &error_suffix) {
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:247]:
```suggestion
                                                         TensorLayout layout, bool pinned,
                                                         AccessOrder order, int device_id,
                                                         const std::string &error_suffix) {
```

File: dali/pipeline/data/tensor_list.cc [hunk 2]
```diff
@@ -273,7 +273,7 @@ void TensorVector<Backend>::VerifySampleShareCompatibility(DALIDataType type, in
 
 
 template <typename Backend>
-void TensorVector<Backend>::UnsafeSetSample(int sample_idx, const TensorVector<Backend> &src,
+void TensorList<Backend>::UnsafeSetSample(int sample_idx, const TensorList<Backend> &src,
                                             int src_sample_idx) {
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:277]:
```suggestion
                                          int src_sample_idx) {
```

File: dali/pipeline/data/tensor_list.cc [hunk 3]
```diff
@@ -322,7 +322,7 @@ void TensorVector<Backend>::UnsafeSetSample(int sample_idx, const Tensor<Backend
 
 
 template <typename Backend>
-void TensorVector<Backend>::UnsafeSetSample(int sample_idx, const shared_ptr<void> &ptr,
+void TensorList<Backend>::UnsafeSetSample(int sample_idx, const shared_ptr<void> &ptr,
                                             size_t bytes, bool pinned, const TensorShape<> &shape,
                                             DALIDataType type, int device_id, AccessOrder order,
                                             const TensorLayout &layout) {
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:328]:
```suggestion
                                          size_t bytes, bool pinned, const TensorShape<> &shape,
                                          DALIDataType type, int device_id, AccessOrder order,
                                          const TensorLayout &layout) {
```

File: dali/pipeline/data/tensor_list.cc [hunk 4]
```diff
@@ -347,7 +347,7 @@ void TensorVector<Backend>::UnsafeSetSample(int sample_idx, const shared_ptr<voi
 
 
 template <typename Backend>
-void TensorVector<Backend>::VerifySampleCopyCompatibility(DALIDataType type, int sample_dim,
+void TensorList<Backend>::VerifySampleCopyCompatibility(DALIDataType type, int sample_dim,
                                                           TensorLayout layout,
                                                           const TensorShape<> &current_shape,
                                                           const TensorShape<> &new_shape,
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:353]:
```suggestion
                                                        TensorLayout layout,
                                                        const TensorShape<> &current_shape,
                                                        const TensorShape<> &new_shape,
```

File: dali/pipeline/data/tensor_list.cc [hunk 5]
```diff
@@ -377,7 +377,7 @@ void TensorVector<Backend>::VerifySampleCopyCompatibility(DALIDataType type, int
 
 
 template <typename Backend>
-void TensorVector<Backend>::UnsafeCopySample(int sample_idx, const TensorVector<Backend> &src,
+void TensorList<Backend>::UnsafeCopySample(int sample_idx, const TensorList<Backend> &src,
                                              int src_sample_idx, AccessOrder order) {
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:381]:
```suggestion
                                           int src_sample_idx, AccessOrder order) {
```

File: dali/pipeline/data/tensor_list.cc [hunk 6]
```diff
@@ -396,7 +396,7 @@ void TensorVector<Backend>::UnsafeCopySample(int sample_idx, const TensorVector<
 
 
 template <typename Backend>
-void TensorVector<Backend>::UnsafeCopySample(int sample_idx, const Tensor<Backend> &src,
+void TensorList<Backend>::UnsafeCopySample(int sample_idx, const Tensor<Backend> &src,
                                              AccessOrder order) {
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:400]:
```suggestion
                                           AccessOrder order) {
```

File: dali/pipeline/data/tensor_list.cc [hunk 7]
```diff
@@ -508,24 +508,24 @@ void TensorVector<Backend>::set_order(AccessOrder order, bool synchronize) {
...
+ConstSampleView<Backend> TensorList<Backend>::operator[](size_t pos) const {
   DALI_ENFORCE(pos < static_cast<size_t>(curr_num_tensors_), "Out of bounds access");
   return {tensors_[pos].raw_data(), shape().tensor_shape_span(pos), tensors_[pos].type()};
 }
 
 
 template <typename Backend>
-void TensorVector<Backend>::Resize(const TensorListShape<> &new_shape, DALIDataType new_type,
+void TensorList<Backend>::Resize(const TensorListShape<> &new_shape, DALIDataType new_type,
                                    BatchContiguity state) {
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:526]:
```suggestion
                                 BatchContiguity state) {
```

File: dali/pipeline/data/tensor_list.cc [hunk 8]
```diff
@@ -977,7 +977,7 @@ Tensor<Backend> TensorVector<Backend>::AsTensor() {
 
 
 template <typename Backend>
-void TensorVector<Backend>::ShareData(const shared_ptr<void> &ptr, size_t bytes, bool pinned,
+void TensorList<Backend>::ShareData(const shared_ptr<void> &ptr, size_t bytes, bool pinned,
                                       const TensorListShape<> &shape, DALIDataType type,
                                       int device_id, AccessOrder order,
                                       const TensorLayout &layout) {
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:983]:
```suggestion
                                    const TensorListShape<> &shape, DALIDataType type,
                                    int device_id, AccessOrder order,
                                    const TensorLayout &layout) {
```

File: dali/pipeline/data/tensor_list.cc [hunk 9]
```diff
@@ -1120,15 +1120,15 @@ bool TensorVector<Backend>::shares_data() const {
 }
 
 
-template class DLL_PUBLIC TensorVector<CPUBackend>;
-template class DLL_PUBLIC TensorVector<GPUBackend>;
-template void TensorVector<CPUBackend>::Copy<CPUBackend>(const TensorVector<CPUBackend> &,
+template class DLL_PUBLIC TensorList<CPUBackend>;
+template class DLL_PUBLIC TensorList<GPUBackend>;
+template void TensorList<CPUBackend>::Copy<CPUBackend>(const TensorList<CPUBackend> &,
                                                          AccessOrder, bool);
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:1126]:
```suggestion
                                                       AccessOrder, bool);
```

File: dali/pipeline/data/tensor_list.cc [hunk 10]
```diff
@@ -1120,15 +1120,15 @@ bool TensorVector<Backend>::shares_data() const {
...
-template void TensorVector<CPUBackend>::Copy<GPUBackend>(const TensorVector<GPUBackend> &,
+template void TensorList<CPUBackend>::Copy<GPUBackend>(const TensorList<GPUBackend> &,
                                                          AccessOrder, bool);
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:1128]:
```suggestion
                                                       AccessOrder, bool);
```

File: dali/pipeline/data/tensor_list.cc [hunk 11]
```diff
@@ -1120,15 +1120,15 @@ bool TensorVector<Backend>::shares_data() const {
...
-template void TensorVector<GPUBackend>::Copy<CPUBackend>(const TensorVector<CPUBackend> &,
+template void TensorList<GPUBackend>::Copy<CPUBackend>(const TensorList<CPUBackend> &,
                                                          AccessOrder, bool);
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:1130]:
```suggestion
                                                       AccessOrder, bool);
```

File: dali/pipeline/data/tensor_list.cc [hunk 12]
```diff
@@ -1120,15 +1120,15 @@ bool TensorVector<Backend>::shares_data() const {
...
-template void TensorVector<GPUBackend>::Copy<GPUBackend>(const TensorVector<GPUBackend> &,
+template void TensorList<GPUBackend>::Copy<GPUBackend>(const TensorList<GPUBackend> &,
                                                          AccessOrder, bool);
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:1132]:
```suggestion
                                                       AccessOrder, bool);
```

**Review by mzient — APPROVED** [review]
Fix the indentation, otherwise OK.

**Comment by klecki** [comment]
!build

**Comment by prak-nv** [comment]
It'd be nice to have commit with old name ``[[deprecated]]`` and aliases for such rename as partial change next time.

**Review by prak-nv — APPROVED** [review]

**Review by awolant — APPROVED** [review]

**Comment by klecki** [comment]
!build

**Pull request merged**